### DISCUSSION TITLE
Example of using cudaStreamAddCallback

### DISCUSSION THREAD

**Discussion by wensimin** [NVIDIA/cuda-python#310] [Q&A] [answered]
I try to register a callback on a stream, and the following code reports an error
```python
def stream_over(*argv):
    print(f"over !")

 cudart.cudaStreamAddCallback(stream, stream_over, None, 0)

```

```
nvidia-dev-1  |   cudart.cudaStreamAddCallback(stream._stream, stream_over, None, 0)
nvidia-dev-1  |   File "cuda/bindings/runtime.pyx", line 15344, in cuda.bindings.runtime.cudaStreamAddCallback
nvidia-dev-1  |   File "cuda/bindings/runtime.pyx", line 4054, in cuda.bindings.runtime.cudaStreamCallback_t.__cinit__
nvidia-dev-1  | TypeError: an integer is required
```

In the documentation I found this class `cudaStreamCallback_t`，but I don't know how to use it correctly  
Any help？

**Comment by leofang** [comment, marked as answer]
Hi @wensimin this is how `cudaStreamAddCallback` (deprecated) and `cudaLaunchHostFunc` are tested:
https://github.com/NVIDIA/cuda-python/blob/eeda1ab5b3b89b1e8ee8ae8dc2f1cbff502b2d5c/cuda_bindings/tests/test_cudart.py#L1293-L1342

**Reply by wensimin** [reply]
I tried to use `cudaStreamAddCallback `and `cudaLaunchHostFunc `according to the examples.
But it seems that the callback will not be executed without using `cudart.cudaDeviceSynchronize()`?
But` cudart.cudaDeviceSynchronize() `is a blocking waiting method. 
How can I get the asynchronous end event of the previous operation in the stream?

**Reply by leofang** [reply]
Hi @wensimin, after the callback is queued on the stream you should be able to
1. call `cudaStreamSynchronize`, or
2. record an event on the stream and then
   a. call `cudaEventSynchronize` , or 
   b. check the `cudaEventQuery` output

to ensure/check if all the works, including the host callback, are executed. CUDA streams are just work queues for the devices, and you won't know if the queue is drained without getting the host involved somehow.
